base/range.jl

```diff
@@ -86,51 +77,140 @@ julia> range(1, 10, length=101)
 
 julia> range(1, 100, step=5)
 1:5:96
+
+julia> range(stop=10, length=5)
+6:10
+
+julia> range(stop=10, step=1, length=5)
+6:1:10
+
+julia> range(start=1, step=1, stop=10)
+1:1:10
 ```
+If `length` is not specified and `stop - start` is not an integer multiple of `step`, a range that ends before `stop` will be produced.
+```jldoctest
+julia> range(1, 3.5, step=2)
+1.0:2.0:3.0
+```
+
+Special care is taken to ensure intermediate values are computed rationally.
+To avoid this induced overhead, see the [`LinRange`](@ref) constructor.
+
+`start` may be specified as either a positional or keyword argument.
+`stop` may be specified as either a positional or keyword argument. If it is specified as a positional argument,
+one of `step` or `length` must also be provided.
+
+!!! compat "Julia 1.1"
+    `stop` as a positional argument requires at least Julia 1.1.
+
+!!! compat "Julia 1.6"
+    `start` as a keyword argument requires at least Julia 1.6.
 """
-range(start; length::Union{Integer,Nothing}=nothing, stop=nothing, step=nothing) =
+function range end
+
+range(start; stop=nothing, length::Union{Integer,Nothing}=nothing, step=nothing) =
     _range(start, step, stop, length)
 
-range(start, stop; length::Union{Integer,Nothing}=nothing, step=nothing) =
-    _range2(start, step, stop, length)
+function range(start, stop; length::Union{Integer,Nothing}=nothing, step=nothing)
+    # For code clarity, the user must pass step or length
+    # See https://github.com/JuliaLang/julia/pull/28708#issuecomment-420034562
+    if step === length === nothing
+        msg = """
+        Neither `step` nor `length` was provided. To fix this do one of the following:
+        * Pass one of them
+        * Use `$(start):$(stop)`
+        * Use `range($start, stop=$stop)`
+        """
+        throw(ArgumentError(msg))
+    end
+    _range(start, step, stop, length)
+end
 
-_range2(start, ::Nothing, stop, ::Nothing) =
-    throw(ArgumentError("At least one of `length` or `step` must be specified"))
+range(;start=nothing, stop=nothing, length::Union{Integer, Nothing}=nothing, step=nothing) =
+    _range(start, step, stop, length)
 
-_range2(start, step, stop, length) = _range(start, step, stop, length)
+_range(start::Nothing, step::Nothing, stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Nothing, stop::Nothing, len::Any    ) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Nothing, stop::Any    , len::Nothing) = range_error(start, step, stop, len)
```

Comment:
```suggestion
_range(start::Nothing, step::Nothing, stop::Any    , len::Nothing) = Base.OneTo(stop)
```
`1`  is a reasonable default for `start`